### PR TITLE
Add multiple endpoints for GMC pipeline via gmcrouter

### PR DESCRIPTION
**Pull request by zhlsunshine**
## Description

Add multiple endpoints for GMC pipeline via gmcrouter.
1. Support new router verb `dataprep` for gmcrouter
2. Support multipart/form-data header for gmcrouter

## Issues

https://github.com/opea-project/GenAIInfra/issues/168
https://github.com/opea-project/GenAIInfra/issues/207
https://github.com/opea-project/GenAIComps/issues/366

## Type of change

List the type of change like below. Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would break existing design and interface)

## Dependencies

`n/a`.

## Tests
**Query before Intel Gaudi information is injected:**
image

**Injecting the Intel Gaudi information**
image

**Query after Intel Gaudi information is injected:**
image
image

**There is no any dataprep router is defined:**
image

### CODE REVIEW

File: microservices-connector/cmd/router/main.go
Comment (irisdingbj) [inline microservices-connector/cmd/router/main.go:486]: There should be only one dataprep service,   so we can break out the loop if we found it

File: microservices-connector/cmd/router/main.go
Comment (irisdingbj) [inline microservices-connector/cmd/router/main.go:534]: add writer into defer() block as well

File: microservices-connector/cmd/router/main.go
Comment (mkbhanda) [reply microservices-connector/cmd/router/main.go:486]: this will get us past the dataprep hump if we break out I like the full pass through in case there are other endpoints. Perhaps multiple different datapreps for different kinds of data - audio, video, images, text, web search, some agent. May be even different embedding services with their own vector database paths.

Also do steps have some description text, so that when someone queries for these endpoints they know what to use for what, like "query path", "data upload path" or something better.

**Review by mkbhanda — COMMENTED** [review]
@irisdingbj  and @zhlsunshine  I think there could be more than two endpoints .. please see my comment. Two is a reasonable default, need to being some context!

File: microservices-connector/cmd/router/main.go
Comment (KfreeZ) [inline microservices-connector/cmd/router/main.go:21]: this is reminding me we should go mod tidy in each release cycle, @irisdingbj ?

File: microservices-connector/cmd/router/main.go
Comment (KfreeZ) [inline microservices-connector/cmd/router/main.go:581]: maybe add a helper here? To indicate the correct path if user input unexpected ones?

**Comment by KfreeZ** [comment]
A question: how we expose the /dataprep in router service's status?

File: microservices-connector/cmd/router/main.go
Comment (zhlsunshine) [reply microservices-connector/cmd/router/main.go:486]: Hi @irisdingbj, I think we need to consider that there are multiple dataprep microservices in Pipeline, such as in switch mode. So I suggest that we can keep this logic, does it make sense?

Comment (zhlsunshine) [reply microservices-connector/cmd/router/main.go:486]: Hi @mkbhanda, the data formats supported by dataprep are only text format, such as txt, htm, md, etc. And for audio, video, images format, especially for video, we may need more time for supporting them because we need to handle the media stream in router. So suggest that we can just support text format as well, does it make sense?

File: microservices-connector/cmd/router/main.go
Comment (zhlsunshine) [reply microservices-connector/cmd/router/main.go:534]: Hi @irisdingbj ,  sure, missing this. Done!

File: microservices-connector/cmd/router/main.go
Comment (zhlsunshine) [reply microservices-connector/cmd/router/main.go:21]: Hi @KfreeZ, will update the go.mod

File: microservices-connector/cmd/router/main.go
Comment (zhlsunshine) [reply microservices-connector/cmd/router/main.go:581]: Hi @KfreeZ, I suggest that add a README for it, does it make sense?

File: microservices-connector/cmd/router/main.go
Comment (zhlsunshine) [reply microservices-connector/cmd/router/main.go:486]: And some additional information, I handle all file in binary format, so in theories, it can handle all these kind of file format, but we need to handle the stream properly.

File: microservices-connector/cmd/router/main.go
Comment (zhlsunshine) [reply microservices-connector/cmd/router/main.go:21]: Hi @KfreeZ, this library is the native lib of golang, so I understand you just want to comment with Iris about this. I thin k we can talk it on meeting.

**Review by KfreeZ — APPROVED** [review]

File: microservices-connector/cmd/router/main.go
Comment (mkbhanda) [reply microservices-connector/cmd/router/main.go:584]: > maybe add a helper here? To indicate the correct path if user input unexpected ones?

@KfreeZ how would we do this? Need to keep all valid paths in some array and if user provides something not in valid list return a usage message?
Do we have a parent class from which our router paths are defined, perhaps we can add this valid paths functionality there. An enhancement feature?

File: microservices-connector/cmd/router/main.go
Comment (mkbhanda) [inline microservices-connector/cmd/router/main.go:584]: To @irisdingbj comment about expecting only one dataprep path and @zhlsunshine allowing/expecting more, is this logic a simplification of adding route only for one dataprep path. We have to be consistent in our logic, support only one now and later more perhaps, or support multiple everywhere.

**Review by mkbhanda — COMMENTED** [review]
Some inline comments. Will watch for you response and then approve.

File: microservices-connector/cmd/router/main.go
Comment (irisdingbj) [reply microservices-connector/cmd/router/main.go:584]: can we add a listService endpoint by default in router? in this way user can list all available endpoint throuth router. because even for dataprep, it will have 3 endpoints, like /dataprep /getDoc /deleteDoc etc.

File: microservices-connector/cmd/router/main.go
Comment (irisdingbj) [reply microservices-connector/cmd/router/main.go:584]: Let's target support more dataprep, in this case, we need generate handler dynamically instead of the fixed /dataprep one.

File: microservices-connector/cmd/router/main.go
Comment (irisdingbj) [inline microservices-connector/cmd/router/main.go:483]: If there is no dataprep in the GMC config, this handler should return info like " this endpoint is not available" to end user.

File: microservices-connector/cmd/router/main.go
Comment (zhlsunshine) [reply microservices-connector/cmd/router/main.go:486]: Yeah, this should be. So I original support the extra request Header of `SERVICE_NAME`, but remove it because I want to make the same format with `dataprep` microservice

File: microservices-connector/cmd/router/main.go
Comment (KfreeZ) [reply microservices-connector/cmd/router/main.go:584]: my thought is like other http routers:
	mux.HandleFunc("/", mcGraphHandler)
	mux.HandleFunc("/dataprep", mcDataHandler)
        mux.HandleFunc("/*other invalid path*", helper) => indicate the legal paths

Comment (KfreeZ) [reply microservices-connector/cmd/router/main.go:584]: according to my search, in the default go router :
    // Catch-all route for handling invalid paths
    router.NoRoute(func(c *gin.Context) {
        c.String(404, "Page not found")
    })

if that works, my thought is maybe we can add a helper here to indicate the correct path to access

File: microservices-connector/cmd/router/main.go
Comment (zhlsunshine) [reply microservices-connector/cmd/router/main.go:485]: Yeah, I added the error feedback when there is no dataprep is found.

File: microservices-connector/cmd/router/main.go
Comment (zhlsunshine) [reply microservices-connector/cmd/router/main.go:598]: Hi @irisdingbj, as we discussed, I will raise an issue for dataprep to DCAI team, they should remove such verb. It's not the best practice for REST API design.

File: microservices-connector/cmd/router/main.go
Comment (zhlsunshine) [reply microservices-connector/cmd/router/main.go:541]: Hi @irisdingbj, after validation, `writer.Close()` should be operated before the request is sent out. So we can not put it in defer() block, otherwise, the request can not contain the input information, such file name or link path, etc.

Comment (zhlsunshine) [reply microservices-connector/cmd/router/main.go:544]: And I also add `handleMultipartError` to handle the error occurred before `writer.Close()`

**Comment by lianhao** [comment]
For the curl (52) error, please wait for PR #249 to land in with the workaround to unblock the issue #258

File: microservices-connector/cmd/router/main.go
Comment (zhlsunshine) [reply microservices-connector/cmd/router/main.go:612]: Hi @irisdingbj, I opened an issue for dataprep REST API definition: https://github.com/opea-project/GenAIComps/issues/394

File: microservices-connector/cmd/router/main.go
Comment (zhlsunshine) [reply microservices-connector/cmd/router/main.go:612]: Hi @KfreeZ, for now, we can just a remind for for user. Like in PR description:
image

**Comment by zhlsunshine** [comment]
Hi @mkbhanda and @irisdingbj, for multiple dataprep REST API definition, I have opened an issue: https://github.com/opea-project/GenAIComps/issues/394 to track it. By following the best practice, there should be only one REST API for the microservice: `/v1/dataprep`, and for all other VectorDB, they really do have only one REST API `/v1/dataprep`. So I think let's see and keep current implementation. And I also add the logic for checking and providing information if there is no `dataprep` router. Does it make sense?

**Comment by zhlsunshine** [comment]
Hi @mkbhanda and @irisdingbj, one more thing is that I remove the request size limitation because there is no size limitation in `dataprep-redis` microserivce now.

**Comment by irisdingbj** [comment]
Let's address listService and dynamic adding endpoints seperately.

**Review by irisdingbj — APPROVED** [review]

**Pull request merged**
